### ISSUE TITLE
[Feature request] Decentralized group management

### ISSUE THREAD

**Issue by MarioSonic** [WhisperSystems/Signal-Android#7150]
<intro>I'm trying to convince the organisation I work for to switch to Signal. One thing that's holding them back (after testing it) is a critical feature: group management. Certainly within a organisation, some kind of group management is crucial. You need to have one or more administrators that have full control over a group. Without it, Signal cannot be used to its full potential.

Crucial example: People who leave to organisation but whose numbers stay active in the group. You cannot expect people to manually leave all groups when they leave an organisation.</intro>

I know Signal doesn't want the server to handle the groups to avoid subpoemas. But maybe there are other ways to implement group management that don't involve centralization. For example that when a group is created, the initiator also generates a group management keypair. All members of the group receive the public key of this keypair. All valid group management messages have to be signed by the group management key. 

When a signed 'removeuser' message is sent to all group members, clients remove that user from the group. An 'adduser' message also has to be signed by and administrator to be accepted by the clients. The administrator of the group can mandate other users of the group by sending them the private key of the group management keypair.

I hope you'll consider adding decentralized group management, since that's one of the last features that gives other messaging apps like WhatsApp an edge and keeps companies from switching to Signal.

**Comment by p4nci** [comment]
related to (not sure if duplicate of) https://github.com/WhisperSystems/Signal-Android/issues/828

https://github.com/WhisperSystems/Signal-Android/issues/828#issuecomment-250123266:

> Example given: a group of colleagues to discuss job related tasks. If one of the group members leaves the team or company, that person should no longer have access to the group,...

https://github.com/WhisperSystems/Signal-Android/issues/828#issuecomment-251792972:

> ...If your previous answer was "decentralized", then our goal should be to first resolve your issue, then implement the decentralized solution.

**Comment by ahirschberg** [comment]
I agree. Having the group administration settings accessible to everyone drastically lowers operational security.

For example, if I created a group of 100 people and set the disappearing message timer to to 1 week, with the expectation that the messages will be securely deleted after that time, and then one out of those 100 people disables the feature, all the messages between then and when someone else catches the mistake and sets the timer back to its proper setting are available forever. This is a privacy issue if any of the 100+ devices is physically compromised at any point in the future.

Having a single or set of group administrator(s) would solve many issues with things like this case, or the case of realizing someone needs to be removed from the group to keep conversation secure or realizing that the group needs to be deleted.

**Comment by Trolldemorted** [comment]
libsignal-protocol-java is perfectly fit for groups with advanced administration methods (and server-side fan-out), WhatsApp is using it and it obviously works.
